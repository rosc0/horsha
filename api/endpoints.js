// Auth
export const TOKEN = '/token'
export const RESET_PASSWORD = '/password_reset/request'
export const REGISTER_DEVICE = '/device'

// Newsfeed
export const NEWSFEED = '/news'
export const JOURNAL_ENTRY = '/journal_entry'
export const COMMENTS = '/comment'
export const LIKE = '/like'

// User
export const USER = '/user'
export const USER_PROFILE = '/user/:id'
export const USER_SEARCH = '/user/search'
export const CURRENT_USER = '/user/me'
export const UPDATE_USER_PROFILE = '/user/:id/profile'
export const UPDATE_USER_PREFERENCES = '/user/:id/preferences'
export const UPDATE_USER_PROFILE_IMAGE = '/user/:id/profile_picture'
export const RESEND_EMAIL_CONFIRMATION =
  '/user/:userId/registered_email/:email/confirmation_state'
export const REGISTERED_EMAIL = '/user/:userId/registered_email'
export const SET_EMAIL_PRIMARY = '/user/:userId/primary_email'
export const CREATE_USER_WITH_FACEBOOK = '/user/facebook'

//Friends
export const REQUEST_FRIENDSHIP = '/user/:user_id/friend_request'
export const ACCEPT_FRIEND_REQUEST =
  '/user/:initiating_user_id/friend_request/:target_user_id/accept'
export const IGNORE_FRIEND_REQUEST =
  '/user/:initiating_user_id/friend_request/:target_user_id/state'
export const DELETE_FRIEND_REQUEST =
  '/user/:initiating_user_id/friend_request/:target_user_id'
export const DELETE_FRIEND = '/user/:user_id/friend/:target_user_id'
export const FRIEND_SUGGESTIONS = '/user/:user_id/friend_suggestions'

// Horses
export const HORSE = '/horse'
export const HORSE_USER = '/horse_user'
export const HORSE_PICTURE = '/profile_picture'
export const HORSE_ALBUM = '/horse_picture'
export const HORSE_USER_RELATION = '/relation'
export const HORSE_JOURNAL = '/journal_entry'
export const HORSE_LAST_ACCESSED = '/horse_user/:id/last_accessed'
export const STATUS_UPDATE = '/status_update'

// Rides
export const RIDE = '/ride'
export const GEO = '/geo/track/gpx'
export const TRACK = '/geo/gpx/track'
export const CONVERT_GPX_WAYLIST = '/geo/gpx/waypoint_list'
// Trails
export const TRAILS = '/trail'
export const TRAIL_PICTURES = '/trail_picture'
export const TRAIL_REVIEW = '/trail_review'
export const TRAIL_POI = '/trail_poi'
export const TRAIL_FAVORITES = '/trail_favorite'
export const TRAIL_RIDES = '/trail_ride'

// Point of Interests
export const POIS = '/poi'

// Report
export const REPORT = '/report'

// Notifications
export const READ_MARKER = '/notification/read_state'
export const NOTIFICATIONS = '/notification'

// Upload
export const UPLOAD = '/upload'
