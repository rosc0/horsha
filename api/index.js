import config from 'react-native-config'
import { build, serialize, emptyToNull } from '@application/utils'
import store from '@application/store'
import TimerMixin from 'react-timer-mixin'
import { setAlertStatus } from '../actions/app'
// import { logoutCurrentUser } from '../actions'
import * as k from '@constants/storageKeys'
import { AsyncStorage } from 'react-native'
// import Authentication from '@api/auth'

const title = 'There’s been a server error'
const subTitle = 'Something’s wrong, please try again later'

// const Auth = new Authentication()

class API {
  constructor(props) {
    this._config = new WeakMap()
    this._endpoint = new WeakMap()
    this._hasBaseURL = false

    if (!props || !props.isAuth) this._setup()
  }

  _setup() {
    const headers = {
      'Content-Type': 'application/json',
    }

    this._config.set(this, { headers })
  }

  setAuthConfig() {
    const headers = {
      'Content-Type': 'application/x-www-form-urlencoded',
    }

    this._config.set(this, { headers })
    return this
  }

  setContentType(contentType) {
    if (contentType) {
      const headers = {
        'Content-Type': contentType,
      }

      this._config.set(this, { headers })
    }
    return this
  }

  _setMethod(method) {
    const config = this._config.get(this) || {}
    config.method = method

    this._config.set(this, config)
  }

  setEndpoint(endpoint, hasBaseURL = false) {
    this._endpoint.set(this, endpoint)

    this._hasBaseURL = hasBaseURL
    return this
  }

  getEndpointPath() {
    return this._endpoint.get(this)
  }

  getEndpoint() {
    if (this._hasBaseURL) {
      return this._endpoint.get(this)
    }

    return config.API_BASE + this._endpoint.get(this)
  }

  getHeaders = () => {
    const state = store.getState()

    const {
      auth: { auth },
      user: { user },
    } = state

    return {
      access_token: auth.accessToken,
      user_id: user.id,
    }
  }

  setBody(body, isJSON) {
    const config = this._config.get(this) || {}

    config.body = isJSON
      ? JSON.stringify(emptyToNull(body))
      : serialize(emptyToNull(body))

    this._config.set(this, config)

    return this
  }

  getConfig() {
    let _config = this._config.get(this)

    const isAuth =
      _config.headers['Content-Type'] === 'application/x-www-form-urlencoded'

    if (isAuth) {
      let isClientIdDuplicated = false

      if (_config.body && !_config.body._parts) {
        isClientIdDuplicated = _config.body.indexOf('client_id') > -1
      }

      if (
        _config.method !== 'GET' &&
        isAuth &&
        !isClientIdDuplicated &&
        !_config.body._parts
      ) {
        _config.body += '&client_id=' + config.CLIENT_ID
      }
    }

    return this._config.get(this)
  }

  _makeRequest() {
    return new Promise((resolve, reject) => {
      let config = this.getConfig()
      let endpoint = this.getEndpoint()

      if (config.method === 'GET') {
        delete config.body
      }

      fetch(endpoint, config)
        .then(response => {
          TimerMixin.setTimeout(() => null, 0)

          if (response.status === 500) {
            // to handler if API return is not valid
            console.warn('throw Error', response)
            store.dispatch(setAlertStatus('error', title, subTitle))
            throw Error(response.statusText)
          }

          if (response.status === 403) {
            console.warn('logged out', response)
            if (
              response._bodyInit !==
              '{"status":403,"description":"User not allowed to allowed to see pictures for horse"}'
            ) {
              AsyncStorage.removeItem(k.AUTH_NEW_TOKENS)

              return store.dispatch({
                type: 'RESET',
                payload: {
                  app: store.getState().app,
                },
              })
            }
            // return response
            AsyncStorage.removeItem(k.AUTH_NEW_TOKENS)

            return store.dispatch({
              type: 'RESET',
              payload: {
                app: store.getState().app,
              },
            })
          }

          if (response.status === 401) {
            if (
              response._bodyInit === '{"code":401,"error":"Invalid token"} '
            ) {
              // ugly workaround
              AsyncStorage.removeItem(k.AUTH_NEW_TOKENS)
              return store.dispatch({
                type: 'RESET',
                payload: {
                  app: store.getState().app,
                },
              })
            }

            return response
          }

          if (response.status === 422) {
            if (
              response._bodyInit === '{"code":422,"text":"Missing user_id"}' ||
              response._bodyInit ===
                '{"status":422,"description":"Invalid UUID string: null"}'
            ) {
              // ugly workaround
              AsyncStorage.removeItem(k.AUTH_NEW_TOKENS)
              return store.dispatch({
                type: 'RESET',
                payload: {
                  app: store.getState().app,
                },
              })
            }

            return response
          }

          if (response.status < 200 || response.status > 299) {
            // to handler if API return is not valid
            console.warn('throw Error', response)
            store.dispatch(setAlertStatus('error', title, subTitle))
            throw Error('Error')
          }

          if (response.status === 204) {
            return response
          }

          const isResponseJSON =
            response.headers.get('content-type') &&
            response.headers.get('content-type').indexOf('application/json') !==
              -1

          return (response.status === 200 || response.status === 201) &&
            isResponseJSON
            ? response.json()
            : response
        })
        .then(res => {
          resolve(res)
        })
        .catch(err => {
          reject(err)
        })

      this._setup()
    })
  }

  addToken(type) {
    const { auth } = store.getState().auth
    const { accessToken: token, expireDate: expires } = auth

    const method =
      type === 'qs'
        ? this.addQueryStringToken.bind(this)
        : this.addJSONToken.bind(this)

    if (token && expires > +new Date()) {
      method(token)
    }
  }

  addQueryStringToken(token) {
    const body = this.getEndpointPath()
    const endpoint = build('get', { body, token })

    this.setEndpoint(endpoint)
  }

  addJSONToken(token) {
    const { body } = this.getConfig()
    const data = build('post', { body, token })

    this._config.set(this, { body: data })
  }

  get(hasToken) {
    if (hasToken) this.addToken('qs')

    this._setMethod('GET')
    this.setBody(null)
    return this._makeRequest()
  }

  post(hasToken) {
    if (hasToken) this.addToken('json')

    this._setMethod('POST')
    return this._makeRequest()
  }

  put(hasToken) {
    if (hasToken) this.addToken('json')

    this._setMethod('PUT')
    return this._makeRequest()
  }

  delete(hasToken, type = 'qs') {
    if (hasToken) this.addToken(type)

    this._setMethod('DELETE')
    return this._makeRequest()
  }
}

export default API
