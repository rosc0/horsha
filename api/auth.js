import { AsyncStorage } from 'react-native'
import * as k from '@constants/storageKeys'
import API from './index'
import * as endpoints from './endpoints'
import { USER_CHANGE_PASSWORD } from '../apollo/mutations/UserCollection'
import { client } from '../apollo/init'

class Authentication extends API {
  constructor({ isAuth = true } = {}) {
    super({ isAuth })
  }

  authenticate(credentials) {
    return new Promise((resolve, reject) => {
      this.setAuthConfig()
        .setEndpoint(endpoints.TOKEN)
        .setBody(credentials)
        .post()
        .then(response =>
          response.access_token ? this.storeToken(response) : response
        )
        .then(resolve)
        .catch(error => {
          this.handleError(error)
          reject()
        })
    })
  }

  refreshToken(refresh_token) {
    const body = {
      grant_type: 'refresh_token',
      refresh_token,
    }

    return new Promise((resolve, reject) => {
      this.setAuthConfig()
        .setEndpoint(endpoints.TOKEN)
        .setBody(body)
        .post()
        .then(this.storeToken)
        .then(resolve)
        .catch(error => {
          this.handleError(error)
          reject()
        })
    })
  }

  async isUserAuthenticated() {
    const authTokenData = await this.getToken()
    if (!authTokenData || !authTokenData.accessToken) {
      console.warn('no token')
      // throw new Error
      this.handleError('no token')
      return false
    }

    if (authTokenData.expireDate > +new Date()) {
      return authTokenData
    }

    return this.refreshToken(authTokenData.refreshToken)
  }

  async getToken() {
    const token = await AsyncStorage.getItem(k.AUTH_NEW_TOKENS)

    if (token) {
      return JSON.parse(token)
    }

    return false
  }

  storeToken(response) {
    const payload = {
      accessToken: response.access_token,
      expireDate: +new Date() + parseInt(response.expires_in) * 1000,
      refreshToken: response.refresh_token,
    }

    try {
      AsyncStorage.setItem(k.AUTH_NEW_TOKENS, JSON.stringify(payload))
    } catch (e) {
      this.handleError(e)
    }

    return payload
  }

  clearStorage() {
    return AsyncStorage.clear()
  }

  async resetPassword(password) {
    const { user_id } = this.getHeaders()

    const a = await client.mutate({
      mutation: USER_CHANGE_PASSWORD,
      variables: {
        password,
        userId: user_id,
      },
    })

    return a.userChangePassword
  }

  handleError(error) {
    console.warn('[AUTH] ERROR. ', error)
  }
}

export default Authentication
