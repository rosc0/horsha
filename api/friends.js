import API from './index'
import * as endpoints from './endpoints'
import { serialize } from '../utils'
import { client } from '../apollo/init'
import {
  GET_PEOPLE_YOU_MAY_KNOW,
  GET_PENDING_INVITATIONS,
  GET_IGNORED_INVITATIONS,
  GET_FRIENDS,
} from '../apollo/queries/UserCollection'
import {
  USER_FRIEND_REQUEST,
  USER_UNFRIEND_REQUEST,
  DENY_FRIEND_REQUEST,
  ACCEPT_FRIEND_REQUEST,
  UPDATE_FRIEND_REQUEST,
  IGNORE_FRIEND_REQUEST,
} from '../apollo/mutations/UserCollection'

class Friends extends API {
  async getFriends(userId = null, cursor = null, order_by = null) {
    const { user_id } = this.getHeaders()

    const {
      data: { userList },
    } = await client.query({
      query: GET_FRIENDS,
      variables: {
        userId: user_id,
        maxItems: 50,
      },
    })

    return userList
  }

  searchFriends(searchTerm, cursor) {
    const { access_token, user_id } = this.getHeaders()

    let params = {
      access_token,
      friends_with_user_id: user_id,
      order_by: '+name',
    }

    if (searchTerm) {
      params.name = searchTerm + '*'
    }

    if (cursor) {
      params.cursor = cursor
    }

    const endpoint = `${endpoints.USER}?${serialize(params)}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  async getInvitations(state = 'pending') {
    const { user_id } = this.getHeaders()

    const {
      data: { userList },
    } = await client.query({
      query:
        state === 'pending' ? GET_PENDING_INVITATIONS : GET_IGNORED_INVITATIONS,
      variables: {
        userId: user_id,
        maxItems: 15,
      },
    })

    return userList
  }

  async getSuggestions() {
    const { user_id } = this.getHeaders()

    const {
      data: { userFriendSuggestions },
    } = await client.query({
      query: GET_PEOPLE_YOU_MAY_KNOW,
      variables: {
        userId: user_id,
        maxItems: 15,
      },
    })

    return userFriendSuggestions
  }

  async acceptRequest(initiatorUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: ACCEPT_FRIEND_REQUEST,
      variables: {
        initiatorUserId,
        userId: user_id,
      },
    })
  }

  async ignoreRequest(initiatorUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: IGNORE_FRIEND_REQUEST,
      variables: {
        initiatorUserId,
        userId: user_id,
      },
    })
  }

  async removeIgnoreRequest(initiatorUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: UPDATE_FRIEND_REQUEST,
      variables: {
        initiatorUserId,
        userId: user_id,
      },
    })
  }

  async deleteRequest(initiatorUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: DENY_FRIEND_REQUEST,
      variables: {
        initiatorUserId,
        userId: user_id,
      },
    })
  }

  async addFriend(targetUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: USER_FRIEND_REQUEST,
      variables: {
        targetUserId,
        userId: user_id,
      },
    })
  }

  async removeFriend(targetUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: USER_UNFRIEND_REQUEST,
      variables: {
        targetUserId,
        userId: user_id,
      },
    })
  }

  async followUser(targetUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: USER_FOLLOW_REQUEST,
      variables: {
        targetUserId,
        userId: user_id,
      },
    })
  }

  async unfollowUser(targetUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: USER_UNFOLLOW_REQUEST,
      variables: {
        targetUserId,
        userId: user_id,
      },
    })
  }

  handleError(err) {
    // console.log('[Friends] ERROR: ', err)
  }
}

export default Friends
