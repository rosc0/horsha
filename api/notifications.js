import { AsyncStorage } from 'react-native'
import API from './index'
import * as endpoints from './endpoints'
import { serialize } from '@application/utils'
import * as k from '@constants/storageKeys'
import { GET_NOTIFICATIONS_COLLECTION } from '../apollo/queries/NotificationsCollection'
import { client } from '../apollo/init'

class Notifications extends API {
  async getUnreadCount() {
    const { user_id } = this.getHeaders()
    // userId
    // const params = {
    //   access_token,
    //   user_id,
    //   is_read: false,
    //   max_items: 0,
    // }

    // const endpoint = `${endpoints.NOTIFICATIONS}?${serialize(params)}`

    // return new Promise((resolve, reject) => {
    //   this.setEndpoint(endpoint)
    //     .get()
    //     .then(resolve)
    //     .catch(reject)
    // })
    // GET_NOTIFICATIONS_COLLECTION
    const data = await client.query({
      query: GET_NOTIFICATIONS_COLLECTION,
      variables: {
        // maxItems,
        // filter: query,
        userId: user_id,
      },
    })
    console.log('@@ notifications query data', data)
    return data
  }

  updateReadMarker(dateRead = null) {
    // const { access_token, user_id } = this.getHeaders()
    // const body = {
    //   user_id,
    //   read_until_date: dateRead ? dateRead : Date.now() * 1000,
    // }
    // const endpoint = `${endpoints.READ_MARKER}?access_token=${access_token}`
    // return new Promise((resolve, reject) => {
    //   this.setEndpoint(endpoint)
    //     .setBody(body, true)
    //     .post()
    //     .then(resolve)
    //     .catch(reject)
    // })
  }

  getNotifications(cursor = null) {
    // const { access_token, user_id } = this.getHeaders()
    // const params = {
    //   access_token,
    //   user_id,
    // }
    // if (cursor) {
    //   params.cursor = cursor
    // }
    // const endpoint = `${endpoints.NOTIFICATIONS}?${serialize(params)}`
    // return new Promise((resolve, reject) => {
    //   this.setEndpoint(endpoint)
    //     .get()
    //     .then(resolve)
    //     .catch(reject)
    // })
  }

  async getNotification(notificationId) {
    const { accessToken } = JSON.parse(
      await AsyncStorage.getItem(k.AUTH_NEW_TOKENS)
    )

    if (accessToken === null) {
      return false
    }

    const params = {
      access_token: accessToken,
    }
    const endpoint = `${endpoints.NOTIFICATIONS}/${notificationId}?${serialize(
      params
    )}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }
}

export default Notifications
