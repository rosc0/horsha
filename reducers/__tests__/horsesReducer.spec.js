import horsesReducer, {
  ADD_HORSE_STEPS,
  initialState as state,
} from '../horses'
import types from '../../constants/actionTypes'

import { horse_picture, horse_user } from '../../../test/api'

describe('Horses Reducer', () => {
  it('should return initial state', () => {
    expect(horsesReducer()).toBe(state)
    expect(horsesReducer()).toMatchSnapshot()
  })

  it('should get horses', () => {
    expect(
      horsesReducer(state, {
        type: types.GET_HORSES_FULFILLED,
        payload: horse_user.GET,
      })
    ).toMatchSnapshot()
  })

  it('should go to previous step', () => {
    expect(
      horsesReducer(
        {
          ...state,
          addHorse: {
            ...state.addHorse,
            step: ADD_HORSE_STEPS.DESCRIPTION,
          },
        },
        {
          type: types.SET_PREVIOUS_STEP,
        }
      )
    ).toMatchSnapshot()
  })

  it('should go to next step', () => {
    expect(
      horsesReducer(
        {
          ...state,
          addHorse: {
            ...state.addHorse,
            step: ADD_HORSE_STEPS.DESCRIPTION,
          },
        },
        {
          type: types.SET_PREVIOUS_STEP,
        }
      )
    ).toMatchSnapshot()
  })

  it('should get horse album', () => {
    expect(
      horsesReducer(state, {
        type: types.GET_HORSE_ALBUM_FULFILLED,
        payload: horse_picture.GET,
      })
    ).toMatchSnapshot()
  })

  it('should get horse user', () => {
    expect(
      horsesReducer(state, {
        type: types.GET_HORSE_USER_FULFILLED,
        payload: horse_user.GET,
      })
    ).toMatchSnapshot()
  })
})
