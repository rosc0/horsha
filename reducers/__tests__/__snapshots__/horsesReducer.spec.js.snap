// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Horses Reducer should get horse album 1`] = `
Object {
  "addHorse": Object {
    "error": false,
    "fetching": false,
    "step": "NAME",
  },
  "album": Object {
    "fetching": false,
    "pictures": Array [
      Object {
        "canDeleteImage": false,
        "height": undefined,
        "image": undefined,
        "key": "461f4f4f-a39d-4cdb-b286-977f9e955195",
        "width": undefined,
      },
      Object {
        "canDeleteImage": false,
        "height": undefined,
        "image": undefined,
        "key": "99ee327c-df20-4bce-9367-8bc986f5b35a",
        "width": undefined,
      },
      Object {
        "canDeleteImage": false,
        "height": undefined,
        "image": undefined,
        "key": "69c5e59c-aa22-46a3-a930-760baec21343",
        "width": undefined,
      },
      Object {
        "canDeleteImage": false,
        "height": undefined,
        "image": undefined,
        "key": "ffb38107-b935-48a9-b35d-584da69c786b",
        "width": undefined,
      },
      Object {
        "canDeleteImage": false,
        "height": undefined,
        "image": undefined,
        "key": "a3e6ea6d-a92c-4581-a715-087b90bab459",
        "width": undefined,
      },
      Object {
        "canDeleteImage": false,
        "height": undefined,
        "image": undefined,
        "key": "59a9d6fb-1255-4a9a-806d-b3b68180e40b",
        "width": undefined,
      },
      Object {
        "canDeleteImage": false,
        "height": undefined,
        "image": undefined,
        "key": "a6da204d-0d81-4142-84ae-55a19b7cba05",
        "width": undefined,
      },
    ],
    "selectedImages": Array [],
    "upload": Object {
      "errored": false,
      "isUploading": false,
      "uploaded": false,
    },
  },
  "cursor": null,
  "detail": Object {
    "entry": null,
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "error": false,
  "fetched": false,
  "fetching": false,
  "forceUpdate": null,
  "horse": null,
  "horsePictureChanged": 0,
  "horseUser": null,
  "horses": Array [],
  "horsesRemaining": 0,
  "journal": Object {
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "profile": Object {
    "birth": null,
    "breed": null,
    "color": null,
    "dateCreated": null,
    "description": null,
    "fetched": false,
    "fetching": true,
    "gender": null,
    "height": null,
    "horseCare": null,
    "image": "",
    "name": null,
    "stats": Object {
      "distance": 0,
      "duration": 0,
      "rides": 0,
    },
    "weight": null,
  },
  "selectedRides": Array [],
  "team": Object {
    "fetching": true,
    "owners": Array [],
    "users": Array [],
  },
}
`;

exports[`Horses Reducer should get horse user 1`] = `
Object {
  "addHorse": Object {
    "error": false,
    "fetching": false,
    "step": "NAME",
  },
  "album": Object {
    "fetching": true,
    "pictures": Array [],
    "selectedImages": Array [],
    "upload": Object {
      "errored": false,
      "isUploading": false,
      "uploaded": false,
    },
  },
  "cursor": null,
  "detail": Object {
    "entry": null,
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "error": false,
  "fetched": false,
  "fetching": false,
  "forceUpdate": null,
  "horse": null,
  "horsePictureChanged": 0,
  "horseUser": Object {
    "date_created": 1465208196000,
    "date_last_accessed": 1501745884000,
    "horse": Object {
      "date_created": 1464792727000,
      "id": "8f13609e-9872-4f90-8deb-3c4dc0a9bd02",
      "profile": Object {
        "name": "wicked long name test hello",
      },
      "profile_picture": Object {
        "fullsize": Object {
          "height": 800,
          "url": "https://dev-content.horsha.com/023f45bb-c202-47cd-8ada-45f66c9167fd.jpg",
          "width": 800,
        },
        "thumbnail": Object {
          "height": 300,
          "url": "https://dev-content.horsha.com/3172cfe9-4b5b-48a4-8dd3-5a10f58c61fc.jpg",
          "width": 300,
        },
      },
    },
    "id": "e97d2af9-1365-44a9-873d-2b55198ea815",
    "relation_type": "sharer",
    "user": Object {
      "id": "95abaa8c-98f6-4eb5-9c84-a7b4abd19d4b",
      "name": "Do",
      "profile_picture": Object {
        "fullsize": Object {
          "height": 1286,
          "url": "https://dev-content.horsha.com/f60b0bb9-7ace-4ee4-959d-4a2a8a219d55.jpg",
          "width": 1600,
        },
        "thumbnail": Object {
          "height": 300,
          "url": "https://dev-content.horsha.com/70d44fd6-6a4f-42d7-a912-04354045defb.jpg",
          "width": 300,
        },
      },
    },
  },
  "horses": Array [],
  "horsesRemaining": 0,
  "journal": Object {
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "profile": Object {
    "birth": null,
    "breed": null,
    "color": null,
    "dateCreated": null,
    "description": null,
    "fetched": false,
    "fetching": true,
    "gender": null,
    "height": null,
    "horseCare": null,
    "image": "",
    "name": null,
    "stats": Object {
      "distance": 0,
      "duration": 0,
      "rides": 0,
    },
    "weight": null,
  },
  "selectedRides": Array [],
  "team": Object {
    "fetching": false,
    "owners": Array [
      false,
    ],
    "users": Array [
      Object {
        "createdAt": 1465208196000,
        "horseUserId": "e97d2af9-1365-44a9-873d-2b55198ea815",
        "image": undefined,
        "key": "95abaa8c-98f6-4eb5-9c84-a7b4abd19d4b",
        "name": "Do",
        "relationType": "sharer",
      },
    ],
  },
}
`;

exports[`Horses Reducer should get horses 1`] = `
Object {
  "addHorse": Object {
    "error": false,
    "fetching": false,
    "step": "NAME",
  },
  "album": Object {
    "fetching": true,
    "pictures": Array [],
    "selectedImages": Array [],
    "upload": Object {
      "errored": false,
      "isUploading": false,
      "uploaded": false,
    },
  },
  "cursor": undefined,
  "detail": Object {
    "entry": null,
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "error": false,
  "fetched": true,
  "fetching": false,
  "forceUpdate": null,
  "horse": Object {
    "date_created": 1464792727000,
    "id": "8f13609e-9872-4f90-8deb-3c4dc0a9bd02",
    "profile": Object {
      "name": "wicked long name test hello",
    },
    "profile_picture": Object {
      "fullsize": Object {
        "height": 800,
        "url": "https://dev-content.horsha.com/023f45bb-c202-47cd-8ada-45f66c9167fd.jpg",
        "width": 800,
      },
      "thumbnail": Object {
        "height": 300,
        "url": "https://dev-content.horsha.com/3172cfe9-4b5b-48a4-8dd3-5a10f58c61fc.jpg",
        "width": 300,
      },
    },
  },
  "horsePictureChanged": 0,
  "horseUser": null,
  "horses": Array [
    Object {
      "date_created": 1465208196000,
      "date_last_accessed": 1501745884000,
      "horse": Object {
        "date_created": 1464792727000,
        "id": "8f13609e-9872-4f90-8deb-3c4dc0a9bd02",
        "profile": Object {
          "name": "wicked long name test hello",
        },
        "profile_picture": Object {
          "fullsize": Object {
            "height": 800,
            "url": "https://dev-content.horsha.com/023f45bb-c202-47cd-8ada-45f66c9167fd.jpg",
            "width": 800,
          },
          "thumbnail": Object {
            "height": 300,
            "url": "https://dev-content.horsha.com/3172cfe9-4b5b-48a4-8dd3-5a10f58c61fc.jpg",
            "width": 300,
          },
        },
      },
      "id": "e97d2af9-1365-44a9-873d-2b55198ea815",
      "relation_type": "sharer",
      "user": Object {
        "id": "95abaa8c-98f6-4eb5-9c84-a7b4abd19d4b",
        "name": "Do",
        "profile_picture": Object {
          "fullsize": Object {
            "height": 1286,
            "url": "https://dev-content.horsha.com/f60b0bb9-7ace-4ee4-959d-4a2a8a219d55.jpg",
            "width": 1600,
          },
          "thumbnail": Object {
            "height": 300,
            "url": "https://dev-content.horsha.com/70d44fd6-6a4f-42d7-a912-04354045defb.jpg",
            "width": 300,
          },
        },
      },
    },
  ],
  "horsesRemaining": 45,
  "journal": Object {
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "profile": Object {
    "birth": null,
    "breed": null,
    "color": null,
    "dateCreated": null,
    "description": null,
    "fetched": false,
    "fetching": true,
    "gender": null,
    "height": null,
    "horseCare": null,
    "image": "",
    "name": null,
    "stats": Object {
      "distance": 0,
      "duration": 0,
      "rides": 0,
    },
    "weight": null,
  },
  "selectedRides": Array [],
  "team": Object {
    "fetching": true,
    "owners": Array [],
    "users": Array [],
  },
}
`;

exports[`Horses Reducer should go to next step 1`] = `
Object {
  "addHorse": Object {
    "error": false,
    "fetching": false,
    "step": "INFO",
  },
  "album": Object {
    "fetching": true,
    "pictures": Array [],
    "selectedImages": Array [],
    "upload": Object {
      "errored": false,
      "isUploading": false,
      "uploaded": false,
    },
  },
  "cursor": null,
  "detail": Object {
    "entry": null,
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "error": false,
  "fetched": false,
  "fetching": false,
  "forceUpdate": null,
  "horse": null,
  "horsePictureChanged": 0,
  "horseUser": null,
  "horses": Array [],
  "horsesRemaining": 0,
  "journal": Object {
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "profile": Object {
    "birth": null,
    "breed": null,
    "color": null,
    "dateCreated": null,
    "description": null,
    "fetched": false,
    "fetching": true,
    "gender": null,
    "height": null,
    "horseCare": null,
    "image": "",
    "name": null,
    "stats": Object {
      "distance": 0,
      "duration": 0,
      "rides": 0,
    },
    "weight": null,
  },
  "selectedRides": Array [],
  "team": Object {
    "fetching": true,
    "owners": Array [],
    "users": Array [],
  },
}
`;

exports[`Horses Reducer should go to previous step 1`] = `
Object {
  "addHorse": Object {
    "error": false,
    "fetching": false,
    "step": "INFO",
  },
  "album": Object {
    "fetching": true,
    "pictures": Array [],
    "selectedImages": Array [],
    "upload": Object {
      "errored": false,
      "isUploading": false,
      "uploaded": false,
    },
  },
  "cursor": null,
  "detail": Object {
    "entry": null,
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "error": false,
  "fetched": false,
  "fetching": false,
  "forceUpdate": null,
  "horse": null,
  "horsePictureChanged": 0,
  "horseUser": null,
  "horses": Array [],
  "horsesRemaining": 0,
  "journal": Object {
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "profile": Object {
    "birth": null,
    "breed": null,
    "color": null,
    "dateCreated": null,
    "description": null,
    "fetched": false,
    "fetching": true,
    "gender": null,
    "height": null,
    "horseCare": null,
    "image": "",
    "name": null,
    "stats": Object {
      "distance": 0,
      "duration": 0,
      "rides": 0,
    },
    "weight": null,
  },
  "selectedRides": Array [],
  "team": Object {
    "fetching": true,
    "owners": Array [],
    "users": Array [],
  },
}
`;

exports[`Horses Reducer should return initial state 1`] = `
Object {
  "addHorse": Object {
    "error": false,
    "fetching": false,
    "step": "NAME",
  },
  "album": Object {
    "fetching": true,
    "pictures": Array [],
    "selectedImages": Array [],
    "upload": Object {
      "errored": false,
      "isUploading": false,
      "uploaded": false,
    },
  },
  "cursor": null,
  "detail": Object {
    "entry": null,
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "error": false,
  "fetched": false,
  "fetching": false,
  "forceUpdate": null,
  "horse": null,
  "horsePictureChanged": 0,
  "horseUser": null,
  "horses": Array [],
  "horsesRemaining": 0,
  "journal": Object {
    "error": false,
    "fetched": false,
    "fetching": false,
  },
  "profile": Object {
    "birth": null,
    "breed": null,
    "color": null,
    "dateCreated": null,
    "description": null,
    "fetched": false,
    "fetching": true,
    "gender": null,
    "height": null,
    "horseCare": null,
    "image": "",
    "name": null,
    "stats": Object {
      "distance": 0,
      "duration": 0,
      "rides": 0,
    },
    "weight": null,
  },
  "selectedRides": Array [],
  "team": Object {
    "fetching": true,
    "owners": Array [],
    "users": Array [],
  },
}
`;
